### PR TITLE
Add scoped typedef to translate-c

### PR DESCRIPTION
**Pull request by matt1795**
Please see #5256. This is my first PR on this project and I'm new to zig so feedback is certainly welcome :)

### CODE REVIEW

File: src-self-hosted/translate_c.zig [hunk 1]
```diff
@@ -1394,6 +1394,24 @@ fn transDeclStmt(rp: RestorePoint, scope: *Scope, stmt: *const ZigClangDeclStmt)
...
+                const underlying_qual = ZigClangTypedefNameDecl_getUnderlyingType(type_decl);
+                const underlying_type = ZigClangQualType_getTypePtr(underlying_qual);
+
+                const mangled_name = try block_scope.makeMangledName(c, name);
+                const node = try transCreateNodeVarDecl(c, false, true, mangled_name);
+                node.eq_token = try appendToken(c, .Equal, "=");
+
+                const loc = ZigClangStmt_getBeginLoc(@ptrCast(*const ZigClangStmt, stmt));
+                node.init_node = try transType(rp, underlying_type, loc);
+                node.semicolon_token = try appendToken(c, .Semicolon, ";");
```
Comment (Vexu) [inline src-self-hosted/translate_c.zig:1412]: Could you extract this into a new function and also make `transTypeDef`.

`transTypeDef` also contains a check for builtin types which this should also be made into a function and called here.

Comment (matt1795) [reply src-self-hosted/translate_c.zig]: Done, please take note of the function names, I'm not sure if they are precise enough.

File: src-self-hosted/translate_c.zig [hunk 2]
```diff
@@ -1394,6 +1405,21 @@ fn transDeclStmt(rp: RestorePoint, scope: *Scope, stmt: *const ZigClangDeclStmt)
...
+            .Typedef => {
+                const typedef_decl = @ptrCast(*const ZigClangTypedefNameDecl, it[0]);
+                const name = try c.str(ZigClangNamedDecl_getName_bytes_begin(
+                    @ptrCast(*const ZigClangNamedDecl, typedef_decl),
+                ));
+
+                const mangled_name = try block_scope.makeMangledName(c, name);
+                if (try checkForBuiltinTypedef(c, typedef_decl, mangled_name)) |ast_node| {
+                    return ast_node;
+                }
```
Comment (Vexu) [inline src-self-hosted/translate_c.zig:1417]: This isn't correct and will cause an assertion failure when rendering this:
```c
int main(int argc, char **argv) {
  typedef int uint32_t;
}
```

To get same behavior as with global builtin typedefs, you should return a string from `checkForBuiltinTypedef` and add it to the list of aliases in `block_scope.variables`.

File: src-self-hosted/translate_c.zig [hunk 3]
```diff
@@ -704,28 +694,49 @@ fn transTypeDef(c: *Context, typedef_decl: *const ZigClangTypedefNameDecl, top_l
...
-    const visib_tok = try appendToken(c, .Keyword_pub, "pub");
-    const const_tok = try appendToken(c, .Keyword_const, "const");
-    const node = try transCreateNodeVarDecl(c, true, true, checked_name);
-    node.eq_token = try appendToken(c, .Equal, "=");
+    const node = (try transZigTypedef(rp, typedef_decl, true, checked_name)) orelse return null;
+    try addTopLevelDecl(c, checked_name, &node.base);
+    return transCreateNodeIdentifier(c, checked_name);
+}
+
+fn transZigTypedef(rp: RestorePoint, typedef_decl: *const ZigClangTypedefNameDecl, toplevel: bool, checked_name: []const u8) Error!?*ast.Node.VarDecl {
```
Comment (Vexu) [inline src-self-hosted/translate_c.zig:724]: I would have gone with something like `transCreateNodeTypedef` to match other similar functions, but that isn't necessarily any better than this.

File: src-self-hosted/translate_c.zig (hunk 2 above)
Comment (matt1795) [reply src-self-hosted/translate_c.zig]: How can you get a string that represents the aliased type? The code I recently pushed doesn't seem to have work as I intended.

**Comment by Vexu** [comment]
Thanks! I added a small patch to get the bulitin type detection working `underlying_type` was the name of the type e.g. `Builtin` for integers.

**Comment by matt1795** [comment]
@Vexu I see now, quick question on the array for checking for builtin types, did you take comptime out because it is redundant there?

**Comment by matt1795** [comment]
And thanks for finishing that one off!

**Comment by Vexu** [comment]
Yes, it's redundant since all the values are constants. If some members were being generated then `comptime` would be needed but only if it's in function scope, container level variables are implicitly `comptime`.

**Pull request merged**